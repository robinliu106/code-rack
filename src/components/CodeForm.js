import React, { useState } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();

const CodeForm = (props) => {
    const [title, setTitle] = useState(props.expense ? props.expense.description : "");
    const [note, setNote] = useState(props.expense ? props.expense.note : "");
    const [createdAt, setCreatedAt] = useState(props.expense ? moment(props.expense.createdAt) : moment());
    const [calendarFocused, setCalendarFocused] = useState(false);
    const [error, setError] = useState("");

    const onTitleChange = (e) => {
        setTitle(e.target.value);
    };
    const onNoteChange = (e) => {
        setNote(e.target.value);
    };

    const onDateChange = (createdAt) => {
        if (createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onFocusChange = ({ focused }) => {
        setCalendarFocused(focused);
    };
    const onSubmit = (e) => {
        e.preventDefault();
        if (!title) {
            setError("Please Provide a title");
        } else {
            setError("");

            props.onSubmit({
                description: title,
                createdAt: createdAt.valueOf(),
                note,
            });
        }
    };

    return (
        <form className="form" onSubmit={onSubmit}>
            {error && <p className="form__error">{error}</p>}
            <input
                className="text-input"
                type="text"
                placeholder="Title"
                autoFocus
                value={title}
                onChange={onTitleChange}
            />

            <SingleDatePicker
                date={createdAt}
                onDateChange={onDateChange}
                focused={calendarFocused}
                onFocusChange={onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
            />
            <textarea className="textarea" value={note} onChange={onNoteChange}></textarea>
            <div>
                <button className="button">Save Code</button>
            </div>
        </form>
    );
};

export default CodeForm;
