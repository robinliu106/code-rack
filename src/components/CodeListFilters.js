import React, { useState } from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";

const CodeListFilters = (props) => {
    const onTextChange = (e) => {
        props.setTextFilter(e.target.value);
    };

    return (
        <div className="content-container">
            <div className="input-group">
                <div className="input-group__item">
                    <input type="text" className="text-input" onChange={onTextChange} placeholder="Search Code" />
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CodeListFilters);
