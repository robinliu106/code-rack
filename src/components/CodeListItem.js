import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const CodeListItem = ({ id, description, note, createdAt }) => (
    <Link className="list-item" to={`/edit/${id}`}>
        <div>
            <h3 className="list-item__title">{description}</h3>
            <span className="list-item__sub-title">{moment(createdAt).format("ddd, MMM YYYY")}</span>
        </div>
        <div>
            <h3 className="list-item__data">{`${note.slice(0, 20)}...`}</h3>
        </div>
    </Link>
);

export default CodeListItem;
