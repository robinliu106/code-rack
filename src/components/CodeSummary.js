import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import selectCodeSnippets from "../selectors/expenses";

export const CodeSummary = ({ codeCount }) => {
    const snippetWord = codeCount === 1 ? "snippet" : "snippets";

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{codeCount}</span> {snippetWord}
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">
                        Add Code
                    </Link>
                </div>
            </div>
        </div>
    );
};
const mapStateToProps = (state) => {
    const visibleSnippets = selectCodeSnippets(state.expenses, state.filters);

    return {
        codeCount: visibleSnippets.length,
    };
};

export default connect(mapStateToProps)(CodeSummary);
