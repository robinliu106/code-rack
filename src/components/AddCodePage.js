import React from "react";
import { connect } from "react-redux";
import CodeForm from "./CodeForm";
import { startAddExpense } from "../actions/expenses";

const AddSnippetPage = (props) => {
    const onSubmit = (expense) => {
        props.startAddExpense(expense);
        props.history.push("/"); //redirect to dashboard upon submit
    };

    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Add Code</h1>
                </div>
            </div>
            <div className="content-container">
                <CodeForm onSubmit={onSubmit} />
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddSnippetPage);
