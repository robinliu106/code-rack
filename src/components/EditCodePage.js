import React from "react";
import { connect } from "react-redux";
import CodeForm from "./CodeForm";
import { startEditSnippet, startRemoveSnippet } from "../actions/expenses";

const EditCodePage = (props) => {
    const onSubmit = (expense) => {
        props.startEditSnippet(props.expense.id, expense);
        props.history.push("/"); //has access to history api because history is passed into the component because the component is registered to a route
    };
    const onRemove = () => {
        props.startRemoveSnippet({ id: props.expense.id });
        props.history.push("/");
    };

    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Code</h1>
                </div>
            </div>
            <div className="content-container">
                <CodeForm expense={props.expense} onSubmit={onSubmit} />

                <button className="button button--secondary" onClick={onRemove}>
                    Remove Expense
                </button>
            </div>
        </div>
    );
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditSnippet: (id, expense) => dispatch(startEditSnippet(id, expense)),
    startRemoveSnippet: (data) => dispatch(startRemoveSnippet(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditCodePage);
