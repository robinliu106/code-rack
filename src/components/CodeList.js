import React from "react";
import { connect } from "react-redux";
import CodeListItem from "./CodeListItem";
import selectSnippets from "../selectors/expenses";

export const CodeList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Snippets</div>
            <div className="show-for-desktop">Snippet</div>
            <div className="show-for-desktop">Description</div>
        </div>
        <div className="list-body">
            {props.expenses.length === 0 ? (
                <div className="list-item list-item--message">
                    <span>No Snippets</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return <CodeListItem key={expense.id} {...expense} />;
                })
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectSnippets(state.expenses, state.filters),
    };
};

export default connect(mapStateToProps)(CodeList);
