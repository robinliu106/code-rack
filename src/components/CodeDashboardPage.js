import React from "react";
import CodeListFilters from "./CodeListFilters";
import CodeSummary from "./CodeSummary";
import CodeList from "./CodeList";

const CodeDashboardPage = () => (
    <div>
        <CodeSummary />
        <CodeListFilters />
        <CodeList />
    </div>
);

export default CodeDashboardPage;
