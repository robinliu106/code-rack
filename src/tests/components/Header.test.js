import React from "react";
import { shallow } from "enzyme";
import { Header } from "../../components/Header";

test("Should render header correctly", () => {
    const wrapper = shallow(<Header startLogout={() => {}} />);

    expect(wrapper.debug()).toMatchSnapshot();
});

test("Should call start logout on button click", () => {
    const startLogout = jest.fn(); //returns a spy
    const wrapper = shallow(<Header startLogout={startLogout} />);
    wrapper.find("button").simulate("click");
    expect(startLogout).toHaveBeenCalled();
});
