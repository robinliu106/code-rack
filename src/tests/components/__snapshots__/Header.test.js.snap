// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render header correctly 1`] = `
"<header>
  <h1>
    Expensify
  </h1>
  <NavLink to=\\"/\\" activeClassName=\\"is-active\\" exact={true}>
    Dashboard
  </NavLink>
  <NavLink to=\\"/create\\" activeClassName=\\"is-active\\">
    Create Expense
  </NavLink>
  <button onClick={[Function]}>
    Logout
  </button>
</header>"
`;
